src/lib/storage/space_maker.rb

```diff
@@ -0,0 +1,76 @@
+#!/usr/bin/env ruby
+#
+# encoding: utf-8
+
+# Copyright (c) [2015] SUSE LLC
+#
+# All Rights Reserved.
+#
+# This program is free software; you can redistribute it and/or modify it
+# under the terms of version 2 of the GNU General Public License as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "fileutils"
+require_relative "./proposal_volume"
+require_relative "./disk_size"
+require_relative "./disk_analyzer"
+require "pp"
+
+module Yast
+  module Storage
+    #
+    # Class to provide free space for creating new partitions - either by
+    # reusing existing unpartitioned space, by deleting existing partitions
+    # or by resizing an existing Windows partition.
+    #
+    class SpaceMaker
+      include Yast::Logger
+
+      attr_reader :volumes
+
+      # Initialize.
+      #
+      # @param volumes [list of ProposalVolume] volumes to find space for.
+      # The volumes might be changed by this class.
+      #
+      # @param settings [Storage::Settings] parameters to use
+      #
+      # @param disk_analyzer [DiskAnalyzer]
+      #
+      def initialize(volumes, settings, disk_analyzer)
```

Comment:
OK, disk analyzer is dependency injection. It is good, as it allows replacement. But it is good habit to provide default, so it is not mandatory to specify it. For me nicer API is to use defaults like
```
  def initialize(volumes, settings: StorageSettings.new, disk_analyzer: DiskAnalyzer.new)

  # and usage
  SpaceMaker.new(volumes)
  SpaceMaker.new(volumes, disk_analyzer: my_own_cool_analyzer)
  SpaceMaker.new(volumes, settings: user_settings || control_file_settings)
```